src/test-run/request/create-request-options.ts

```diff
@@ -173,16 +169,31 @@ function getProxyUrl (testRun: TestRun, url: string, withCredentials?: boolean):
     }, testRun, true)) as Promise<string>;
 }
 
+async function resolveUrlParts (testRun: TestRun, url: URL, withCredentials: boolean): Promise<{ hostname: string; port: string; href: string; partAfterHost: string }> {
```

Comment:
I suggest creating two methods: `resolveProxyUrlParts` and `resolveProxylessUrlParts`. Then call them into this method depending on `testRun.isProxyless()`. It will be prettier in my opinion. But it's up to you, it's not a strict recommendation.